File: server/vance-shared/src/main/java/de/mhus/vance/shared/document/kind/ChartCodec.java
package de.mhus.vance.shared.document.kind;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.jspecify.annotations.Nullable;
import tools.jackson.core.JacksonException;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;

/**
 * Parser and serialiser for {@code kind: chart} document bodies —
 * mirrors {@code chartCodec.ts}. JSON and YAML only; markdown is
 * intentionally not supported.
 *
 * <p>One discriminator ({@link ChartType}) covers all chart kinds; the
 * per-type data-point shape is validated on read, with malformed
 * points silently dropped and empty series elided (see
 * {@code specification/doc-kind-chart.md} §2.5 + §3.1).
 *
 * <p>Each data point survives round-trip in the form it arrived in —
 * an object-form point ({@code {x: …, y: …}}) is re-emitted as an
 * object, a tuple-form point ({@code [x, y]}) as a tuple. The codec
 * does not normalise between the two.
 */
public final class ChartCodec {

    private static final ObjectMapper JSON = new ObjectMapper();
    private static final TypeReference<LinkedHashMap<String, Object>> JSON_MAP =
            new TypeReference<>() {};

    private ChartCodec() {
        // utility class
    }

    public static ChartDocument parse(String body, @Nullable String mimeType) {
        if (isJson(mimeType)) return parseJson(body);
        if (isYaml(mimeType)) return parseYaml(body);
        throw new KindCodecException("Unsupported mime type for chart: " + mimeType);
    }

    public static String serialize(ChartDocument doc, @Nullable String mimeType) {
        if (isJson(mimeType)) return serializeJson(doc);
        if (isYaml(mimeType)) return serializeYaml(doc);
        throw new KindCodecException("Unsupported mime type for chart: " + mimeType);
    }

    public static boolean supports(@Nullable String mimeType) {
        return isJson(mimeType) || isYaml(mimeType);
    }

    // ── Mime ───────────────────────────────────────────────────────

    private static boolean isJson(@Nullable String mime) {
        return "application/json".equals(mime);
    }

    private static boolean isYaml(@Nullable String mime) {
        return "application/yaml".equals(mime)
                || "application/x-yaml".equals(mime)
                || "text/yaml".equals(mime)
                || "text/x-yaml".equals(mime);
    }

    // ── JSON / YAML ────────────────────────────────────────────────

    private static ChartDocument parseJson(String body) {
        if (body.isBlank()) return ChartDocument.empty();
        Map<String, Object> parsed;
        try {
            parsed = JSON.readValue(body, JSON_MAP);
        } catch (JacksonException e) {
            throw new KindCodecException("Invalid JSON: " + e.getOriginalMessage(), e);
        }
        if (parsed == null) throw new KindCodecException("Top-level JSON must be an object");
        return promoteToDocument(KindHeaderCodec.unwrapJsonMeta(parsed));
    }

    private static ChartDocument parseYaml(String body) {
        if (body.isBlank()) return ChartDocument.empty();
        return promoteToDocument(KindHeaderCodec.parseYamlBody(body));
    }

    private static String serializeJson(ChartDocument doc) {
        Map<String, Object> wrapped = KindHeaderCodec.wrapJsonMeta(canonicalKind(doc), buildBody(doc));
        try {
            return JSON.writerWithDefaultPrettyPrinter().writeValueAsString(wrapped) + "\n";
        } catch (JacksonException e) {
            throw new KindCodecException("Failed to write JSON: " + e.getOriginalMessage(), e);
        }
    }

    private static String serializeYaml(ChartDocument doc) {
        return KindHeaderCodec.dumpYamlBody(canonicalKind(doc), buildBody(doc));
    }

    // ── Promotion ──────────────────────────────────────────────────

    private static ChartDocument promoteToDocument(Map<String, Object> obj) {
        Object kindRaw = obj.get("kind");
        String kind = (kindRaw instanceof String s) ? s : "";

        ChartHeader header = promoteHeader(obj.get("chart"));
        ChartType type = header.chartType();

        // Axes are silently dropped for pie / donut (no axis semantics)
        // — keeping them around would round-trip noise the spec calls
        // out as "ignored" in §2.5.
        ChartAxis xAxis = type.isNamedValueShaped()
                ? null
                : promoteAxis(obj.get("xAxis"), AxisType.CATEGORY);
        ChartAxis yAxis = type.isNamedValueShaped()
                ? null
                : promoteAxis(obj.get("yAxis"), AxisType.VALUE);

        List<ChartSeries> series = promoteSeries(obj.get("series"), type);
        Map<String, Object> override = promoteOverride(obj.get("echartsOptionOverride"));

        Map<String, Object> extra = new LinkedHashMap<>(obj);
        extra.remove("kind");
        extra.remove("chart");
        extra.remove("xAxis");
        extra.remove("yAxis");
        extra.remove("series");
        extra.remove("echartsOptionOverride");

        return new ChartDocument(
                kind.isEmpty() ? "chart" : kind,
                header, xAxis, yAxis, series, override, extra);
    }

    private static ChartHeader promoteHeader(@Nullable Object raw) {
        if (!(raw instanceof Map<?, ?> map)) {
            throw new KindCodecException("Missing or non-object `chart` block");
        }
        Object typeRaw = map.get("chartType");
        if (!(typeRaw instanceof String typeStr)) {
            throw new KindCodecException("Missing or non-string `chart.chartType`");
        }
        ChartType type = ChartType.fromWire(typeStr);
        if (type == null) {
            throw new KindCodecException("Unknown chartType: " + typeStr);
        }
        String title = (map.get("title") instanceof String s1) ? s1 : null;
        String subtitle = (map.get("subtitle") instanceof String s2) ? s2 : null;
        boolean legend = !(map.get("legend") instanceof Boolean lb) || lb;
        boolean stacked = map.get("stacked") instanceof Boolean sb && sb;
        boolean smooth = map.get("smooth") instanceof Boolean smb && smb;
        return new ChartHeader(type, title, subtitle, legend, stacked, smooth);
    }

    private static ChartAxis promoteAxis(@Nullable Object raw, AxisType defaultType) {
        if (!(raw instanceof Map<?, ?> map)) {
            // No axis block on disk → default axis for this side.
            return new ChartAxis(defaultType, null, null, null, new ArrayList<>());
        }
        AxisType type = defaultType;
        if (map.get("type") instanceof String ts) {
            AxisType parsed = AxisType.fromWire(ts);
            if (parsed != null) type = parsed;
        }
        String label = (map.get("label") instanceof String ls) ? ls : null;
        Double min = (map.get("min") instanceof Number mn) ? mn.doubleValue() : null;
        Double max = (map.get("max") instanceof Number mx) ? mx.doubleValue() : null;
        List<String> categories = new ArrayList<>();
        if (map.get("categories") instanceof List<?> catList) {
            for (Object c : catList) {
                if (c instanceof String cs) categories.add(cs);
                else if (c instanceof Number cn) categories.add(cn.toString());
            }
        }
        return new ChartAxis(type, label, min, max, categories);
    }

    private static List<ChartSeries> promoteSeries(@Nullable Object raw, ChartType type) {
        List<ChartSeries> out = new ArrayList<>();
        if (!(raw instanceof List<?> list)) return out;
        for (Object r : list) {
            if (!(r instanceof Map<?, ?> map)) continue;
            Object nameRaw = map.get("name");
            if (!(nameRaw instanceof String name) || name.isBlank()) continue;
            String color = (map.get("color") instanceof String cs && !cs.isEmpty()) ? cs : null;

            List<Object> data = new ArrayList<>();
            if (map.get("data") instanceof List<?> dataList) {
                for (Object pt : dataList) {
                    if (isValidDataPoint(pt, type)) data.add(pt);
                }
            }
            // Series with zero valid points after shape-filtering are
            // elided — they'd render as an empty legend entry with no
            // bars/lines, which is just noise.
            if (data.isEmpty()) continue;

            Map<String, Object> extra = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                if (!(e.getKey() instanceof String key)) continue;
                if ("name".equals(key) || "color".equals(key) || "data".equals(key)) continue;
                extra.put(key, e.getValue());
            }
            out.add(new ChartSeries(name, color, data, extra));
        }
        return out;
    }

    @SuppressWarnings("unchecked")
    private static @Nullable Map<String, Object> promoteOverride(@Nullable Object raw) {
        if (!(raw instanceof Map<?, ?> map)) return null;
        Map<String, Object> out = new LinkedHashMap<>();
        for (Map.Entry<?, ?> e : map.entrySet()) {
            if (e.getKey() instanceof String k) out.put(k, e.getValue());
        }
        return out.isEmpty() ? null : out;
    }

    /**
     * Per-point shape validation. Tuple-form ({@code List}) requires the
     * minimum number of elements; object-form ({@code Map}) requires the
     * keys the renderer needs to make sense of the point. Wrong types
     * inside (e.g. {@code y} as a string) survive — the renderer will
     * coerce or drop them; the codec is only the structural gate.
     */
    private static boolean isValidDataPoint(@Nullable Object point, ChartType type) {
        if (point instanceof Map<?, ?> map) {
            return switch (type) {
                case LINE, BAR, AREA, SCATTER -> map.containsKey("x") && map.containsKey("y");
                case PIE, DONUT -> map.containsKey("name") && map.containsKey("value");
                case CANDLESTICK -> map.containsKey("t") && map.containsKey("o")
                        && map.containsKey("h") && map.containsKey("l") && map.containsKey("c");
                case HEATMAP -> map.containsKey("x") && map.containsKey("y") && map.containsKey("v");
            };
        }
        if (point instanceof List<?> list) {
            // pie/donut don't have a tuple form — there's no canonical
            // ordering between name and value when both could be either
            // string or number depending on the chart.
            return switch (type) {
                case LINE, BAR, AREA, SCATTER -> list.size() >= 2;
                case PIE, DONUT -> false;
                case CANDLESTICK -> list.size() >= 5;
                case HEATMAP -> list.size() >= 3;
            };
        }
        return false;
    }

    // ── Body builder ───────────────────────────────────────────────

    private static Map<String, Object> buildBody(ChartDocument doc) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("chart", headerToMap(doc.chart()));
        if (doc.xAxis() != null) body.put("xAxis", axisToMap(doc.xAxis()));
        if (doc.yAxis() != null) body.put("yAxis", axisToMap(doc.yAxis()));
        body.put("series", seriesToList(doc.series()));
        if (doc.echartsOptionOverride() != null && !doc.echartsOptionOverride().isEmpty()) {
            body.put("echartsOptionOverride", doc.echartsOptionOverride());
        }
        for (Map.Entry<String, Object> e : doc.extra().entrySet()) {
            if (!body.containsKey(e.getKey())) body.put(e.getKey(), e.getValue());
        }
        return body;
    }

    private static Map<String, Object> headerToMap(ChartHeader h) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("chartType", h.chartType().wire());
        if (h.title() != null) m.put("title", h.title());
        if (h.subtitle() != null) m.put("subtitle", h.subtitle());
        // legend defaults to true — emit only when explicitly false so
        // the on-disk shape stays compact.
        if (!h.legend()) m.put("legend", false);
        if (h.stacked()) m.put("stacked", true);
        if (h.smooth()) m.put("smooth", true);
        return m;
    }

    private static Map<String, Object> axisToMap(ChartAxis a) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("type", a.type().wire());
        if (a.label() != null) m.put("label", a.label());
        if (a.min() != null) m.put("min", a.min());
        if (a.max() != null) m.put("max", a.max());
        if (!a.categories().isEmpty()) m.put("categories", new ArrayList<>(a.categories()));
        return m;
    }

    private static List<Map<String, Object>> seriesToList(List<ChartSeries> series) {
        List<Map<String, Object>> out = new ArrayList<>(series.size());
        for (ChartSeries s : series) {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("name", s.name());
            if (s.color() != null) m.put("color", s.color());
            m.put("data", new ArrayList<>(s.data()));
            for (Map.Entry<String, Object> ex : s.extra().entrySet()) {
                if (!m.containsKey(ex.getKey())) m.put(ex.getKey(), ex.getValue());
            }
            out.add(m);
        }
        return out;
    }

    private static String canonicalKind(ChartDocument doc) {
        return (doc.kind() == null || doc.kind().isBlank()) ? "chart" : doc.kind();
    }
}

File: client/packages/vance-face/src/document/chartCodec.ts
// Codec for `kind: chart` documents — parses an on-disk body into a
// typed ChartDocument and serializes it back. JSON and YAML only;
// markdown is intentionally not supported (see
// `specification/doc-kind-chart.md` §3.3).
//
// One discriminator (chartType) covers all chart kinds; the per-type
// data-point shape is validated on read, with malformed points
// silently dropped and empty series elided.
//
// Each data point survives round-trip in the form it arrived in — an
// object-form point ({x, y}) is re-emitted as an object, a tuple-form
// point ([x, y]) as a tuple. The codec does not normalise between the
// two.

import {
  dumpYamlBody,
  parseYamlBody,
  unwrapJsonMeta,
  wrapJsonMeta,
} from '@vance/shared';

export type ChartType =
  | 'line'
  | 'bar'
  | 'area'
  | 'scatter'
  | 'pie'
  | 'donut'
  | 'candlestick'
  | 'heatmap';

export type AxisType = 'category' | 'value' | 'time' | 'log';

const CHART_TYPES: ReadonlySet<ChartType> = new Set([
  'line', 'bar', 'area', 'scatter', 'pie', 'donut', 'candlestick', 'heatmap',
]);

const AXIS_TYPES: ReadonlySet<AxisType> = new Set([
  'category', 'value', 'time', 'log',
]);

export interface ChartHeader {
  chartType: ChartType;
  title?: string;
  subtitle?: string;
  /** Whether the legend is shown. Default true. */
  legend: boolean;
  /** Whether series are stacked (bar / area / line only). Default false. */
  stacked: boolean;
  /** Whether line / area series use spline interpolation. Default false. */
  smooth: boolean;
}

export interface ChartAxis {
  type: AxisType;
  label?: string;
  min?: number;
  max?: number;
  /** Explicit category order; only meaningful when {@code type} is
   *  {@code category}. Empty by default. */
  categories: string[];
}

/** A single data point. Object-form (Map<string, unknown>) or tuple-
 *  form (unknown[]); the renderer interprets the shape based on the
 *  document's {@code chartType}. */
export type ChartDataPoint = Record<string, unknown> | unknown[];

export interface ChartSeries {
  name: string;
  color?: string;
  data: ChartDataPoint[];
  /** Unknown per-series fields, preserved across round-trip. */
  extra: Record<string, unknown>;
}

export interface ChartDocument {
  /** Always `'chart'`. */
  kind: string;
  chart: ChartHeader;
  /** {@code null} for pie / donut where axes are meaningless. */
  xAxis: ChartAxis | null;
  yAxis: ChartAxis | null;
  series: ChartSeries[];
  /** Raw ECharts option subtree, deep-merged over the schema-generated
   *  option at render time. {@code null} when not present on disk. */
  echartsOptionOverride: Record<string, unknown> | null;
  /** Unknown top-level fields. Re-emitted verbatim on save. */
  extra: Record<string, unknown>;
}

export class ChartCodecError extends Error {
  constructor(message: string, public override readonly cause?: unknown) {
    super(message);
    this.name = 'ChartCodecError';
  }
}

// ── MIME helpers ─────────────────────────────────────────────────────

function isJson(mime: string): boolean {
  return mime === 'application/json';
}

function isYaml(mime: string): boolean {
  return mime === 'application/yaml'
    || mime === 'application/x-yaml'
    || mime === 'text/yaml'
    || mime === 'text/x-yaml';
}

// ── Public API ───────────────────────────────────────────────────────

export function parseChart(body: string, mimeType: string): ChartDocument {
  if (isJson(mimeType)) return parseChartJson(body);
  if (isYaml(mimeType)) return parseChartYaml(body);
  throw new ChartCodecError(`Unsupported mime type for chart: ${mimeType}`);
}

export function serializeChart(doc: ChartDocument, mimeType: string): string {
  if (isJson(mimeType)) return serializeChartJson(doc);
  if (isYaml(mimeType)) return serializeChartYaml(doc);
  throw new ChartCodecError(`Unsupported mime type for chart: ${mimeType}`);
}

/** Whether the codec can handle this mime type — used by the editor
 *  to decide whether to offer the Chart tab. Markdown is intentionally
 *  excluded; spec §3.3. */
export function isChartMime(mimeType: string | null | undefined): boolean {
  if (!mimeType) return false;
  return isJson(mimeType) || isYaml(mimeType);
}

/** {@code true} when the chart type uses {x, y}-shaped data points
 *  (line, bar, area, scatter). Useful for type-switch compatibility
 *  checks in the editor. */
export function isXyShaped(type: ChartType): boolean {
  return type === 'line' || type === 'bar' || type === 'area' || type === 'scatter';
}

/** {@code true} for pie / donut — data points are {name, value}. */
export function isNamedValueShaped(type: ChartType): boolean {
  return type === 'pie' || type === 'donut';
}

// ── JSON ─────────────────────────────────────────────────────────────

function parseChartJson(body: string): ChartDocument {
  if (body.trim() === '') return emptyDoc();
  let parsed: unknown;
  try {
    parsed = JSON.parse(body);
  } catch (e) {
    throw new ChartCodecError('Invalid JSON: ' + (e instanceof Error ? e.message : String(e)), e);
  }
  if (!isObject(parsed)) {
    throw new ChartCodecError('Top-level JSON must be an object');
  }
  return promoteToChartDocument(unwrapJsonMeta(parsed));
}

function serializeChartJson(doc: ChartDocument): string {
  return JSON.stringify(wrapJsonMeta(doc.kind || 'chart', buildOnDiskBody(doc)), null, 2) + '\n';
}

// ── YAML ─────────────────────────────────────────────────────────────

function parseChartYaml(body: string): ChartDocument {
  if (body.trim() === '') return emptyDoc();
  let merged: Record<string, unknown>;
  try {
    merged = parseYamlBody(body);
  } catch (e) {
    throw new ChartCodecError('Invalid YAML: ' + (e instanceof Error ? e.message : String(e)), e);
  }
  return promoteToChartDocument(merged);
}

function serializeChartYaml(doc: ChartDocument): string {
  return dumpYamlBody(doc.kind || 'chart', buildOnDiskBody(doc));
}

// ── Shared promotion + writeback ────────────────────────────────────

export function emptyDoc(): ChartDocument {
  return {
    kind: 'chart',
    chart: {
      chartType: 'line',
      legend: true,
      stacked: false,
      smooth: false,
    },
    xAxis: { type: 'category', categories: [] },
    yAxis: { type: 'value', categories: [] },
    series: [],
    echartsOptionOverride: null,
    extra: {},
  };
}

function promoteToChartDocument(obj: Record<string, unknown>): ChartDocument {
  const kind = typeof obj.kind === 'string' ? obj.kind : 'chart';
  const chart = promoteHeader(obj.chart);

  // Pie / donut have no axis semantics — drop any axis blocks that
  // accidentally made it onto disk.
  const xAxis = isNamedValueShaped(chart.chartType)
    ? null
    : promoteAxis(obj.xAxis, 'category');
  const yAxis = isNamedValueShaped(chart.chartType)
    ? null
    : promoteAxis(obj.yAxis, 'value');

  const series = promoteSeries(obj.series, chart.chartType);
  const echartsOptionOverride = promoteOverride(obj.echartsOptionOverride);

  const {
    kind: _k,
    chart: _c,
    xAxis: _x,
    yAxis: _y,
    series: _s,
    echartsOptionOverride: _o,
    ...extra
  } = obj;

  return {
    kind: kind || 'chart',
    chart,
    xAxis,
    yAxis,
    series,
    echartsOptionOverride,
    extra,
  };
}

function promoteHeader(raw: unknown): ChartHeader {
  if (!isObject(raw)) {
    throw new ChartCodecError('Missing or non-object `chart` block');
  }
  const typeRaw = raw.chartType;
  if (typeof typeRaw !== 'string') {
    throw new ChartCodecError('Missing or non-string `chart.chartType`');
  }
  const type = typeRaw.trim().toLowerCase() as ChartType;
  if (!CHART_TYPES.has(type)) {
    throw new ChartCodecError(`Unknown chartType: ${typeRaw}`);
  }
  const title = typeof raw.title === 'string' ? raw.title : undefined;
  const subtitle = typeof raw.subtitle === 'string' ? raw.subtitle : undefined;
  // legend defaults to true; we only flip it false when explicitly set.
  const legend = typeof raw.legend === 'boolean' ? raw.legend : true;
  const stacked = typeof raw.stacked === 'boolean' ? raw.stacked : false;
  const smooth = typeof raw.smooth === 'boolean' ? raw.smooth : false;
  return { chartType: type, title, subtitle, legend, stacked, smooth };
}

function promoteAxis(raw: unknown, defaultType: AxisType): ChartAxis {
  if (!isObject(raw)) {
    return { type: defaultType, categories: [] };
  }
  let type = defaultType;
  if (typeof raw.type === 'string') {
    const candidate = raw.type.trim().toLowerCase() as AxisType;
    if (AXIS_TYPES.has(candidate)) type = candidate;
  }
  const label = typeof raw.label === 'string' ? raw.label : undefined;
  const min = typeof raw.min === 'number' && Number.isFinite(raw.min) ? raw.min : undefined;
  const max = typeof raw.max === 'number' && Number.isFinite(raw.max) ? raw.max : undefined;
  const categories: string[] = [];
  if (Array.isArray(raw.categories)) {
    for (const c of raw.categories) {
      if (typeof c === 'string') categories.push(c);
      else if (typeof c === 'number' && Number.isFinite(c)) categories.push(String(c));
    }
  }
  return { type, label, min, max, categories };
}

function promoteSeries(raw: unknown, chartType: ChartType): ChartSeries[] {
  if (!Array.isArray(raw)) return [];
  const out: ChartSeries[] = [];
  for (const r of raw) {
    if (!isObject(r)) continue;
    const nameRaw = r.name;
    if (typeof nameRaw !== 'string' || !nameRaw.trim()) continue;
    const color = typeof r.color === 'string' && r.color ? r.color : undefined;

    const data: ChartDataPoint[] = [];
    if (Array.isArray(r.data)) {
      for (const pt of r.data) {
        if (isValidDataPoint(pt, chartType)) data.push(pt as ChartDataPoint);
      }
    }
    // Series with zero valid points are elided — they'd render as an
    // empty legend entry with no bars/lines.
    if (data.length === 0) continue;

    const extra: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(r)) {
      if (k === 'name' || k === 'color' || k === 'data') continue;
      extra[k] = v;
    }
    out.push({ name: nameRaw, color, data, extra });
  }
  // If the on-disk body offered series entries but none survived
  // validation, the input is structurally wrong (most often raw
  // ECharts options with `series[].type` and no `name`/`data`).
  // A silent empty chart would hide the mistake; throw so the editor
  // surfaces the parse error and the LLM sees the contract on retry.
  if (out.length === 0 && raw.length > 0) {
    throw new ChartCodecError(
      `No valid series in \`series\` (input had ${raw.length} entries). `
      + 'Each series needs `name` (string) and `data` (non-empty array of points '
      + `matching chartType \`${chartType}\`, e.g. ${sampleShape(chartType)}). `
      + 'Vance charts use this schema directly — do not use raw ECharts options '
      + 'like `dataset` or bare `series[].type`.',
    );
  }
  return out;
}

function sampleShape(type: ChartType): string {
  switch (type) {
    case 'line':
    case 'bar':
    case 'area':
    case 'scatter':
      return "{ x: 'A', y: 10 }";
    case 'pie':
    case 'donut':
      return "{ name: 'A', value: 10 }";
    case 'candlestick':
      return "{ t: '2024-01-01', o: 1, h: 2, l: 0.5, c: 1.5 }";
    case 'heatmap':
      return '{ x: 0, y: 0, v: 1 }';
  }
}

function promoteOverride(raw: unknown): Record<string, unknown> | null {
  if (!isObject(raw)) return null;
  const keys = Object.keys(raw);
  if (keys.length === 0) return null;
  return raw;
}

/** Per-point shape validation. Tuple-form (Array) requires the minimum
 *  element count; object-form (Record) requires the keys the renderer
 *  needs. Wrong inner types (e.g. {@code y} as a string) survive — the
 *  renderer coerces or drops them. The codec is only the structural
 *  gate. */
function isValidDataPoint(pt: unknown, type: ChartType): boolean {
  if (isObject(pt)) {
    switch (type) {
      case 'line':
      case 'bar':
      case 'area':
      case 'scatter':
        return 'x' in pt && 'y' in pt;
      case 'pie':
      case 'donut':
        return 'name' in pt && 'value' in pt;
      case 'candlestick':
        return 't' in pt && 'o' in pt && 'h' in pt && 'l' in pt && 'c' in pt;
      case 'heatmap':
        return 'x' in pt && 'y' in pt && 'v' in pt;
    }
  }
  if (Array.isArray(pt)) {
    switch (type) {
      case 'line':
      case 'bar':
      case 'area':
      case 'scatter':
        return pt.length >= 2;
      case 'pie':
      case 'donut':
        // No canonical ordering for {name, value} when both can be of
        // different types depending on the chart context.
        return false;
      case 'candlestick':
        return pt.length >= 5;
      case 'heatmap':
        return pt.length >= 3;
    }
  }
  return false;
}

// ── On-disk body builder ────────────────────────────────────────────

function buildOnDiskBody(doc: ChartDocument): Record<string, unknown> {
  const body: Record<string, unknown> = {
    chart: headerToObject(doc.chart),
  };
  if (doc.xAxis) body.xAxis = axisToObject(doc.xAxis);
  if (doc.yAxis) body.yAxis = axisToObject(doc.yAxis);
  body.series = doc.series.map(seriesToObject);
  if (doc.echartsOptionOverride && Object.keys(doc.echartsOptionOverride).length > 0) {
    body.echartsOptionOverride = doc.echartsOptionOverride;
  }
  for (const [k, v] of Object.entries(doc.extra)) {
    if (!(k in body)) body[k] = v;
  }
  return body;
}

function headerToObject(h: ChartHeader): Record<string, unknown> {
  const o: Record<string, unknown> = { chartType: h.chartType };
  if (h.title !== undefined) o.title = h.title;
  if (h.subtitle !== undefined) o.subtitle = h.subtitle;
  // legend defaults to true — only emit when explicitly false so the
  // on-disk shape stays compact.
  if (!h.legend) o.legend = false;
  if (h.stacked) o.stacked = true;
  if (h.smooth) o.smooth = true;
  return o;
}

function axisToObject(a: ChartAxis): Record<string, unknown> {
  const o: Record<string, unknown> = { type: a.type };
  if (a.label !== undefined) o.label = a.label;
  if (a.min !== undefined) o.min = a.min;
  if (a.max !== undefined) o.max = a.max;
  if (a.categories.length > 0) o.categories = [...a.categories];
  return o;
}

function seriesToObject(s: ChartSeries): Record<string, unknown> {
  const o: Record<string, unknown> = { name: s.name };
  if (s.color !== undefined) o.color = s.color;
  o.data = [...s.data];
  for (const [k, v] of Object.entries(s.extra)) {
    if (!(k in o)) o[k] = v;
  }
  return o;
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}
